File: tests/algorithms/transpiler/FS/ciphers/morse_code.fs
// Generated 2025-08-07 10:31 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let CHARS: string array = [|"A"; "B"; "C"; "D"; "E"; "F"; "G"; "H"; "I"; "J"; "K"; "L"; "M"; "N"; "O"; "P"; "Q"; "R"; "S"; "T"; "U"; "V"; "W"; "X"; "Y"; "Z"; "1"; "2"; "3"; "4"; "5"; "6"; "7"; "8"; "9"; "0"; "&"; "@"; ":"; ","; "."; "'"; "\""; "?"; "/"; "="; "+"; "-"; "("; ")"; "!"; " "|]
let CODES: string array = [|".-"; "-..."; "-.-."; "-.."; "."; "..-."; "--."; "...."; ".."; ".---"; "-.-"; ".-.."; "--"; "-."; "---"; ".--."; "--.-"; ".-."; "..."; "-"; "..-"; "...-"; ".--"; "-..-"; "-.--"; "--.."; ".----"; "..---"; "...--"; "....-"; "....."; "-...."; "--..."; "---.."; "----."; "-----"; ".-..."; ".--.-."; "---..."; "--..--"; ".-.-.-"; ".----."; ".-..-."; "..--.."; "-..-."; "-...-"; ".-.-."; "-....-"; "-.--."; "-.--.-"; "-.-.--"; "/"|]
let rec to_upper_char (c: string) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable c = c
    try
        if c = "a" then
            __ret <- "A"
            raise Return
        if c = "b" then
            __ret <- "B"
            raise Return
        if c = "c" then
            __ret <- "C"
            raise Return
        if c = "d" then
            __ret <- "D"
            raise Return
        if c = "e" then
            __ret <- "E"
            raise Return
        if c = "f" then
            __ret <- "F"
            raise Return
        if c = "g" then
            __ret <- "G"
            raise Return
        if c = "h" then
            __ret <- "H"
            raise Return
        if c = "i" then
            __ret <- "I"
            raise Return
        if c = "j" then
            __ret <- "J"
            raise Return
        if c = "k" then
            __ret <- "K"
            raise Return
        if c = "l" then
            __ret <- "L"
            raise Return
        if c = "m" then
            __ret <- "M"
            raise Return
        if c = "n" then
            __ret <- "N"
            raise Return
        if c = "o" then
            __ret <- "O"
            raise Return
        if c = "p" then
            __ret <- "P"
            raise Return
        if c = "q" then
            __ret <- "Q"
            raise Return
        if c = "r" then
            __ret <- "R"
            raise Return
        if c = "s" then
            __ret <- "S"
            raise Return
        if c = "t" then
            __ret <- "T"
            raise Return
        if c = "u" then
            __ret <- "U"
            raise Return
        if c = "v" then
            __ret <- "V"
            raise Return
        if c = "w" then
            __ret <- "W"
            raise Return
        if c = "x" then
            __ret <- "X"
            raise Return
        if c = "y" then
            __ret <- "Y"
            raise Return
        if c = "z" then
            __ret <- "Z"
            raise Return
        __ret <- c
        raise Return
        __ret
    with
        | Return -> __ret
let rec to_upper (s: string) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable s = s
    try
        let mutable res: string = ""
        let mutable i: int = 0
        while i < (String.length (s)) do
            res <- res + (to_upper_char (string (s.[i])))
            i <- i + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
let rec index_of (xs: string array) (target: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable xs = xs
    let mutable target = target
    try
        let mutable i: int = 0
        while i < (Seq.length (xs)) do
            if (_idx xs (i)) = target then
                __ret <- i
                raise Return
            i <- i + 1
        __ret <- -1
        raise Return
        __ret
    with
        | Return -> __ret
let rec encrypt (message: string) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable message = message
    try
        let msg: string = to_upper (message)
        let mutable res: string = ""
        let mutable i: int = 0
        while i < (String.length (msg)) do
            let c: string = string (msg.[i])
            let idx: int = index_of (CHARS) (c)
            if idx >= 0 then
                if res <> "" then
                    res <- res + " "
                res <- res + (_idx CODES (idx))
            i <- i + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
let rec split_spaces (s: string) =
    let mutable __ret : string array = Unchecked.defaultof<string array>
    let mutable s = s
    try
        let mutable res: string array = [||]
        let mutable current: string = ""
        let mutable i: int = 0
        while i < (String.length (s)) do
            let ch: string = string (s.[i])
            if ch = " " then
                if current <> "" then
                    res <- Array.append res [|current|]
                    current <- ""
            else
                current <- current + ch
            i <- i + 1
        if current <> "" then
            res <- Array.append res [|current|]
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
let rec decrypt (message: string) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable message = message
    try
        let parts: string array = split_spaces (message)
        let mutable res: string = ""
        for code in Seq.map string (parts) do
            let idx: int = index_of (CODES) (code)
            if idx >= 0 then
                res <- res + (_idx CHARS (idx))
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
let msg: string = "Morse code here!"
printfn "%s" (msg)
let enc: string = encrypt (msg)
printfn "%s" (enc)
let dec: string = decrypt (enc)
printfn "%s" (dec)
printfn "%s" (encrypt ("Sos!"))
printfn "%s" (decrypt ("... --- ... -.-.--"))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

File: tests/algorithms/x/PHP/ciphers/morse_code.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
$__start_mem = memory_get_usage();
$__start = _now();
  $CHARS = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '&', '@', ':', ',', '.', '\'', '"', '?', '/', '=', '+', '-', '(', ')', '!', ' '];
  $CODES = ['.-', '-...', '-.-.', '-..', '.', '..-.', '--.', '....', '..', '.---', '-.-', '.-..', '--', '-.', '---', '.--.', '--.-', '.-.', '...', '-', '..-', '...-', '.--', '-..-', '-.--', '--..', '.----', '..---', '...--', '....-', '.....', '-....', '--...', '---..', '----.', '-----', '.-...', '.--.-.', '---...', '--..--', '.-.-.-', '.----.', '.-..-.', '..--..', '-..-.', '-...-', '.-.-.', '-....-', '-.--.', '-.--.-', '-.-.--', '/'];
  function to_upper_char($c) {
  global $CHARS, $CODES, $msg, $enc, $dec;
  if ($c == 'a') {
  return 'A';
}
  if ($c == 'b') {
  return 'B';
}
  if ($c == 'c') {
  return 'C';
}
  if ($c == 'd') {
  return 'D';
}
  if ($c == 'e') {
  return 'E';
}
  if ($c == 'f') {
  return 'F';
}
  if ($c == 'g') {
  return 'G';
}
  if ($c == 'h') {
  return 'H';
}
  if ($c == 'i') {
  return 'I';
}
  if ($c == 'j') {
  return 'J';
}
  if ($c == 'k') {
  return 'K';
}
  if ($c == 'l') {
  return 'L';
}
  if ($c == 'm') {
  return 'M';
}
  if ($c == 'n') {
  return 'N';
}
  if ($c == 'o') {
  return 'O';
}
  if ($c == 'p') {
  return 'P';
}
  if ($c == 'q') {
  return 'Q';
}
  if ($c == 'r') {
  return 'R';
}
  if ($c == 's') {
  return 'S';
}
  if ($c == 't') {
  return 'T';
}
  if ($c == 'u') {
  return 'U';
}
  if ($c == 'v') {
  return 'V';
}
  if ($c == 'w') {
  return 'W';
}
  if ($c == 'x') {
  return 'X';
}
  if ($c == 'y') {
  return 'Y';
}
  if ($c == 'z') {
  return 'Z';
}
  return $c;
};
  function to_upper($s) {
  global $CHARS, $CODES, $msg, $enc, $dec;
  $res = '';
  $i = 0;
  while ($i < strlen($s)) {
  $res = $res . to_upper_char(substr($s, $i, $i + 1 - $i));
  $i = $i + 1;
};
  return $res;
};
  function index_of($xs, $target) {
  global $CHARS, $CODES, $msg, $enc, $dec;
  $i = 0;
  while ($i < count($xs)) {
  if ($xs[$i] == $target) {
  return $i;
}
  $i = $i + 1;
};
  return -1;
};
  function encrypt($message) {
  global $CHARS, $CODES, $enc, $dec;
  $msg = to_upper($message);
  $res = '';
  $i = 0;
  while ($i < strlen($msg)) {
  $c = substr($msg, $i, $i + 1 - $i);
  $idx = index_of($CHARS, $c);
  if ($idx >= 0) {
  if ($res != '') {
  $res = $res . ' ';
};
  $res = $res . $CODES[$idx];
}
  $i = $i + 1;
};
  return $res;
};
  function split_spaces($s) {
  global $CHARS, $CODES, $msg, $enc, $dec;
  $res = [];
  $current = '';
  $i = 0;
  while ($i < strlen($s)) {
  $ch = substr($s, $i, $i + 1 - $i);
  if ($ch == ' ') {
  if ($current != '') {
  $res = _append($res, $current);
  $current = '';
};
} else {
  $current = $current . $ch;
}
  $i = $i + 1;
};
  if ($current != '') {
  $res = _append($res, $current);
}
  return $res;
};
  function decrypt($message) {
  global $CHARS, $CODES, $msg, $enc, $dec;
  $parts = split_spaces($message);
  $res = '';
  foreach ($parts as $code) {
  $idx = index_of($CODES, $code);
  if ($idx >= 0) {
  $res = $res . $CHARS[$idx];
}
};
  return $res;
};
  $msg = 'Morse code here!';
  echo rtrim($msg), PHP_EOL;
  $enc = encrypt($msg);
  echo rtrim($enc), PHP_EOL;
  $dec = decrypt($enc);
  echo rtrim($dec), PHP_EOL;
  echo rtrim(encrypt('Sos!')), PHP_EOL;
  echo rtrim(decrypt('... --- ... -.-.--')), PHP_EOL;
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;
